### PR TITLE
feat(core): load environment variables from configuration name

### LINKED ISSUE #17686: Add .[configuration-name].env to environment variables
- [x] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description

Given https://nx.dev/recipes/environment-variables/define-environment-variables

Add

```
1. apps/my-app/.env.[target-name].[configuration-name]
2. apps/my-app/.[target-name].[configuration-name].env
3. apps/my-app/.env.[target-name]
4. apps/my-app/.[target-name].env
5. apps/my-app/.env.local
6. apps/my-app/.local.env
7. apps/my-app/.env
8. .env.[target-name].[configuration-name]
9. .[target-name].[configuration-name].env
10. .env.[target-name]
11. .[target-name].env

--> .env.[configuration-name]
--> [configuration-name].env

12. .local.env
13. .env.local
14. .env
```

## Motivation

Allows a fallback env for any stage, such as shared test credentials. Otherwise we would have to create a separate env in each lib's folder for the configuration noame

## Suggested Implementation

Will create PR

### PR DESCRIPTION
**Pull request by AgentEnder**
Adds support for loading environment variables from .env.[configurationName] and .[configurationName].env

Fixes: #17686

### CODE REVIEW

**Comment by SmoshySmosh** [comment]
This solution does not provide the proper documentation and it is not clear how to use it without having deeper knowledge of the internals of NX.

I originally explored this as a quick and easy solution but I quickly realized it was untenable due to the synchronization overhead it would add to a mono repo of any size. This change requires a re-definition of every configuration in every target of every package in your mono repo, which you then need to keep in sync as you add new packages. We should think through this problem a little more and find a solution that both decreases the complexity of NX and serves the need of NX's customers.

For example, this is what every target will have to become to support `.beta.env`, `.gamma.env`, and `.prod.env` 


```
"targets: {
  ...,
  "e2e": {
    ...
    "configurations": {
      "beta": {},
      "gamma": {},
      "prod": {},
    }
  }
}
```

File: packages/nx/src/utils/project-graph-utils.ts [hunk 1]
```diff
@@ -22,9 +23,28 @@ export function projectHasTargetAndConfiguration(
...
-    project.data.targets[target].configurations &&
-    project.data.targets[target].configurations[configuration]
+    // Explicitly defined target + configuration
+    (projectHasTarget(project, target) &&
+      project.data.targets[target].configurations &&
+      project.data.targets[target].configurations[configuration]) ||
+    // Inferred configuration from presence of .env files
+    (projectHasTarget(project, target) &&
+      [
+        join(workspaceRoot, `.env.${configuration}`),
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/project-graph-utils.ts:33]: I don't think having a file there should define a configuration. Configurations should still be defined in the project configuration. If we'd like an easier way to define configurations, that is a separate discussion.

Can we remove this part from the changes here and consider them separately?

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

File: e2e/nx-run/src/run.test.ts [hunk 2]
```diff
@@ -198,6 +198,15 @@ describe('Nx Running Tests', () => {
...
           },
+          nx: {
+            targets: {
+              [target]: {
+                configurations: {
+                  production: {},
+                },
+              },
+            },
+          },
```
Comment (FrozenPandaz) [inline e2e/nx-run/src/run.test.ts:209]: Were you going to do something with this?

Comment (AgentEnder) [reply e2e/nx-run/src/run.test.ts:209]: Its required since we don't infer the production configuration from .env.production existing anymore

Comment (AgentEnder) [reply e2e/nx-run/src/run.test.ts:209]: We have to explicitly list that configuration if we want to be able to read env variables from the file / run with --configuration production

Comment (AgentEnder) [reply e2e/nx-run/src/run.test.ts:209]: I was, yes. It looks like the e2e I had added somehow got lost 🤔

File: packages/nx/src/utils/project-graph-utils.ts (hunk 1 above)
Comment (SmoshySmosh) [reply packages/nx/src/utils/project-graph-utils.ts]: > I don't think having a file there should define a configuration. Configurations should still be defined in the project configuration. If we'd like an easier way to define configurations, that is a separate discussion.

Removing the requirement to redefine a configuration under every target is what myself and other customers are trying to get. With the current state of environment files in NX, a great deal of technical debt and configuration overhead is generated making it an unusable solution for most projects ultimately making NX very limiting.

**Review by FrozenPandaz — APPROVED** [review]

**Pull request merged**